src/shared/components/Projects/ProjectForm.tsx

```diff
@@ -0,0 +1,264 @@
+import { boolean } from '@storybook/addon-knobs';
+
+/**
+ * Custom from controls, based on:
+ * https://ant.design/components/form/#components-form-demo-customized-form-controls
+ */
+export interface PrefixMappingGroupInputState {
+  prefix: string;
+  namespace: string;
+}
+export interface PrefixMappingGroupInputProps {
+  groupId: string | number;
+  value: any;
+  onChange(state: PrefixMappingGroupInputState): void;
+}
+class PrefixMappingGroupInput extends React.Component<
+  PrefixMappingGroupInputProps,
+  PrefixMappingGroupInputState
+> {
+  // Static method called by Form component
+  static getDerivedStateFromProps(nextProps: PrefixMappingGroupInputProps) {
+    if ('value' in nextProps) {
+      return {
+        ...(nextProps.value || {}),
+      };
+    }
+    return null;
+  }
+
+  constructor(props: PrefixMappingGroupInputProps) {
+    super(props);
+    const value = props.value || {};
+    this.state = {
+      prefix: value.prefix || '',
+      namespace: value.namespace || '',
+    };
+  }
+
+  handlePrefixChange = (e: React.FormEvent<HTMLInputElement>) => {
+    const prefix = e.currentTarget.value;
+    this.setState({
+      prefix,
+    });
+    this.props.onChange({ ...this.state, prefix });
+  };
+
+  handleNamespaceChange = (e: React.FormEvent<HTMLInputElement>) => {
+    const namespace = e.currentTarget.value;
+    this.setState({
+      namespace,
+    });
+    this.props.onChange({ ...this.state, namespace });
+  };
+
+  render() {
+    return (
+      <Input.Group
+        compact
+        style={{ width: 'calc(100% - 22px)', marginRight: 8 }}
+      >
+        <Input
+          style={{ width: '20%' }}
+          name={`prefix[${this.props.groupId}]`}
+          placeholder="prefix"
+          onChange={this.handlePrefixChange}
+          value={this.state.prefix}
+        />
+        <Input
+          style={{ width: '80%' }}
+          name={`namespace[${this.props.groupId}]`}
+          placeholder="namespace"
+          onChange={this.handleNamespaceChange}
+          value={this.state.namespace}
+        />
+      </Input.Group>
+    );
+  }
+}
+
+export interface ProjectFormProps {
+  form: WrappedFormUtils;
+  project?: {
+    label: string;
+    name: string;
+    base?: string;
+    prefixMappings?: PrefixMappingGroupInputState[];
+  };
+  busy?: boolean;
+  onSubmit?(project: ProjectFormProps['project']): any;
+}
+
+/**
+ * Adaptation of the following example:
+ * based on: https://ant.design/components/form/#components-form-demo-dynamic-form-item
+ */
+const ProjectForm: React.FunctionComponent<ProjectFormProps> = ({
+  form,
+  project,
+  busy = false,
+  onSubmit = () => {},
+}) => {
+  // logic for generating dynamic prefix mapping fields in form
+  const currentId =
+    project && project.prefixMappings ? project.prefixMappings.length : 0;
+  const activeKeys =
+    currentId === 0
+      ? []
+      : Array(currentId - 1 + 1)
+          .fill(0)
+          .map((_, idx) => 1 + idx);
```

Comment:
Simpler version of the whole expression, including the ternary operator:
```javascript
[...(Array(currentId + 1).keys())].slice(1);
``` 

Works if currentId === 0. Will return 1-indexed values up to currentId

Reply:
:+1:

Reply:
done in d5c49ea0abac17ca076043eea6f1316c5e523407